test/e2e/firewall.go

```diff
@@ -0,0 +1,410 @@
+	Expect(err).NotTo(HaveOccurred(), "error waiting for daemon pod to be reaped")
+}
+
+func getNonMaterNodeNames(c clientset.Interface) (sets.String, error) {
+	masterName := framework.TestContext.CloudConfig.MasterName
+	nodeList, err := c.Core().Nodes().List(api.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	nodeNames := sets.NewString()
+	for _, node := range nodeList.Items {
+		if node.Name != masterName {
+			nodeNames.Insert(node.Name)
+		}
+	}
+	return nodeNames, nil
+}
+
+func examineHitNodesFromOutside(externalIP string, httpPort int32, timeout time.Duration, expectedHosts sets.String, exitWhenSatisfy bool) error {
+	hittedHosts := sets.NewString()
+
+	framework.Logf("Waiting up to %v for satisfying expectedHosts", timeout)
+	condition := func() (bool, error) {
+		respBody, err := queryFromUrl(externalIP, "hostname", httpPort, firewallTestHttpTimeout)
+		if err != nil {
+			framework.Logf("%v", err)
+			return false, nil
+		}
+		hittedHost := strings.TrimSpace(respBody)
+		hittedHosts.Insert(hittedHost)
+		if !expectedHosts.Has(hittedHost) {
+			return false, fmt.Errorf("error hitting unexpected host: %v", hittedHost)
+		}
+
+		framework.Logf("Missing %+v, got %+v", expectedHosts.Difference(hittedHosts), hittedHosts)
+		if exitWhenSatisfy && hittedHosts.Equal(expectedHosts) {
+			return true, nil
+		}
+		return false, nil
+	}
+
+	if err := wait.Poll(time.Second, timeout, condition); err != nil {
+		if !exitWhenSatisfy && hittedHosts.Equal(expectedHosts) {
+			return nil
+		}
+		return fmt.Errorf("error waiting for expectedHosts: %v, hittedHosts: %v, err message: %v", expectedHosts, hittedHosts, err)
+	}
+	return nil
+}
+
+func examineNotHitNodeFromOutside(externalIP string, httpPort int32, timeout time.Duration, targetHost string, countToSucceed int) error {
+	count := 0
+
+	framework.Logf("Waiting up to %v for not hitting %v for %v times", timeout, targetHost, countToSucceed)
+	condition := func() (bool, error) {
+		respBody, err := queryFromUrl(externalIP, "hostname", httpPort, firewallTestHttpTimeout)
+		if err != nil {
+			framework.Logf("%v", err)
+			return false, nil
+		}
+		hittedHost := strings.TrimSpace(respBody)
+		if hittedHost == targetHost {
+			framework.Logf("Error hitting unexpected host: %v, reset counter: %v", targetHost, count)
+			count = 0
+		} else {
+			count++
+		}
+
+		if count >= countToSucceed {
+			return true, nil
+		}
+		return false, nil
+	}
+
+	if err := wait.Poll(time.Second, timeout, condition); err != nil {
+		return fmt.Errorf("error: not hitting %v for %v times, expect at least %v times, err message: %v", targetHost, count, countToSucceed, err)
+	}
+	return nil
+}
+
+func queryFromUrl(ip, path string, port int32, timeout time.Duration) (string, error) {
+	// Create a new Transport for not re-using connnections
+	client := &http.Client{
+		Timeout:   timeout,
+		Transport: &http.Transport{},
+	}
+	url := fmt.Sprintf("http://%v:%v/%v", ip, port, path)
```

Comment:
there's probably already a function that does this (maybe it's called simpleGET?). If we don't have such a function, suggest adding it to framework since it's really common.